libkbfs/bare_root_metadata_v3.go

```diff
@@ -0,0 +1,986 @@
+
+package libkbfs
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/keybase/client/go/protocol/keybase1"
+	"github.com/keybase/go-codec/codec"
+)
+
+// WriterMetadataV3 stores the metadata for a TLF that is
+// only editable by users with writer permissions.
+type WriterMetadataV3 struct {
+	// Serialized, possibly encrypted, version of the PrivateMetadata
+	SerializedPrivateMetadata []byte `codec:"data"`
+
+	// The last KB user with writer permissions to this TLF
+	// who modified this WriterMetadata
+	LastModifyingWriter keybase1.UID `codec:"lmw"`
+
+	// For public TLFs (since those don't have any keys at all).
+	Writers []keybase1.UID `codec:",omitempty"`
+	// Writers identified by unresolved social assertions.
+	UnresolvedWriters []keybase1.SocialAssertion `codec:"uw,omitempty"`
+	// Pointer to the writer key bundle for private TLFs.
+	WKeyBundleID TLFWriterKeyBundleID `codec:"wkid,omitempty"`
+	// Latest key generation.
+	LatestKeyGen KeyGen `codec:"lkg"`
+
+	// The directory ID, signed over to make verification easier
+	ID TlfID
+	// The branch ID, currently only set if this is in unmerged per-device history.
+	BID BranchID
+	// Flags
+	WFlags WriterFlags
+
+	// Estimated disk usage at this revision
+	DiskUsage uint64
+	// The total number of bytes in new blocks
+	RefBytes uint64
+	// The total number of bytes in unreferenced blocks
+	UnrefBytes uint64
+
+	codec.UnknownFieldSetHandler
+}
+
+// BareRootMetadataV3 is the MD that is signed by the reader or
+// writer. Unlike RootMetadata, it contains exactly the serializable
+// metadata.
+type BareRootMetadataV3 struct {
+	// The metadata that is only editable by the writer.
+	WriterMetadata WriterMetadataV3 `codec:"wmd"`
+	// The signature for the writer metadata, to prove
+	// that it's only been changed by writers.
+	WriterMetadataSigInfo SignatureInfo `codec:"wmdsi"`
+
+	// The last KB user who modified this BareRootMetadata
+	LastModifyingUser keybase1.UID
+	// Flags
+	Flags MetadataFlags
+	// The revision number
+	Revision MetadataRevision
+	// Pointer to the previous root block ID
+	PrevRoot MdID
+
+	// For private TLFs. Any unresolved social assertions for readers.
+	UnresolvedReaders []keybase1.SocialAssertion `codec:"ur,omitempty"`
+	// Pointer to the reader key bundle for private TLFs.
+	RKeyBundleID TLFReaderKeyBundleID `codec:"rkid,omitempty"`
+
+	// ConflictInfo is set if there's a conflict for the given folder's
+	// handle after a social assertion resolution.
+	ConflictInfo *TlfHandleExtension `codec:"ci,omitempty"`
+	// FinalizedInfo is set if there are no more valid writer keys capable
+	// of writing to the given folder.
+	FinalizedInfo *TlfHandleExtension `codec:"fi,omitempty"`
+
+	codec.UnknownFieldSetHandler
+}
+
+// ExtraMetadataV3 contains references to key bundles stored outside of metadata
+// blocks.  This only ever exists in memory and is never serialized itself.
+type ExtraMetadataV3 struct {
+	wkb *TLFWriterKeyBundleV3
+	rkb *TLFReaderKeyBundleV2
```

Comment:
Was avoiding creating the new type since there's no functional change but it will probably reduce confusion. Will add.